spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/dao/MapTaskExecutionDao.java

```diff
@@ -244,11 +249,74 @@ public int compare(TaskExecution e1, TaskExecution e2) {
 		});
 	}
 
-	private Page getPageFromList(List<TaskExecution> executionList, Pageable pageable, long maxSize){
+	private Page<TaskExecution> getPageFromList(List<TaskExecution> executionList, Pageable pageable, long maxSize){
 		long toIndex = (pageable.getOffset() + pageable.getPageSize() > executionList.size()) ?
 				executionList.size() : pageable.getOffset() + pageable.getPageSize();
 		return new PageImpl<>(
 				executionList.subList((int)pageable.getOffset(), (int)toIndex),
 				pageable, maxSize);
 	}
+
+	@Override
+	public List<TaskExecution> getLatestTaskExecutionsByTaskNames(String... taskNames) {
+
+		Assert.notEmpty(taskNames, "At least 1 task name must be provided.");
+
+		final List<String> taskNamesAsList = Arrays.stream(taskNames)
+				.filter(s -> StringUtils.hasText(s)).collect(Collectors.toList());
+
+		Assert.isTrue(taskNamesAsList.size() == taskNames.length,
+			String.format("Task names must not contain any empty elements but %s of %s were empty or null.",
+					taskNames.length - taskNamesAsList.size(), taskNames.length));
+
+		final Map<String, TaskExecution> tempTaskExecutions = new TreeMap<>();
+
+		for (Map.Entry<Long, TaskExecution> taskExecutionMapEntry : this.taskExecutions.entrySet()) {
+			if (!taskNamesAsList.contains(taskExecutionMapEntry.getValue().getTaskName())) {
+				continue;
+			}
+
+			final TaskExecution tempTaskExecution = tempTaskExecutions.get(taskExecutionMapEntry.getValue().getTaskName());
+			if (tempTaskExecution == null
+				|| tempTaskExecution.getStartTime().before(taskExecutionMapEntry.getValue().getStartTime())
+				|| (
+					tempTaskExecution.getStartTime().equals(taskExecutionMapEntry.getValue().getStartTime())
+						&& tempTaskExecution.getExecutionId() < taskExecutionMapEntry.getValue().getExecutionId()
+					)
+				) {
+				tempTaskExecutions.put(taskExecutionMapEntry.getValue().getTaskName(), taskExecutionMapEntry.getValue());
+			}
+		}
+		final List<TaskExecution> latestTaskExecutions = new ArrayList<>(tempTaskExecutions.values());
+		Collections.sort(latestTaskExecutions, new TaskExecutionComparator());
```

Comment:
I still have the question about why we need a sort here if we are using a `TreeeMap` above.

Reply:
@mminella The comparator for a `TreeMap` only sorts the keys … but the `TaskExecutionComparator` sorts on multiple properties in the `TaskExecution`.  The latest commit changes from a `Treemap` to a `HashMap` as the initial sorting is not needed.